pkg/core/server/pos.go

```diff
@@ -0,0 +1,391 @@
+package server
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/AudiusProject/audiusd/pkg/core/common"
+	"github.com/AudiusProject/audiusd/pkg/core/db"
+	"github.com/AudiusProject/audiusd/pkg/core/gen/core_proto"
+	"github.com/AudiusProject/audiusd/pkg/pos"
+	"github.com/cometbft/cometbft/crypto/ed25519"
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgtype"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	mediorumPoSRequestTimeout = 1 * time.Second
+	posChallengeDeadline      = 2
+	posVerificationDelay      = posChallengeDeadline * 3 * time.Second
+)
+
+// Called during FinalizeBlock. Keeps Proof of Storage subsystem up to date with current block.
+func (s *Server) syncPoS(ctx context.Context, latestBlockHash []byte, latestBlockHeight int64) error {
+	if blockShouldTriggerNewPoSChallenge(latestBlockHash) {
+		s.logger.Info("PoS Challenge triggered", "height", latestBlockHeight, "hash", hex.EncodeToString(latestBlockHash))
+		err := s.db.InsertPoSChallenge(ctx, latestBlockHeight)
+		if err != nil {
+			return fmt.Errorf("Could not insert PoS challenge to db at height %d: %v", latestBlockHeight, err)
+		}
+		// TODO: disable in block sync mode
+		go s.sendPoSChallengeToMediorum(latestBlockHash, latestBlockHeight)
+	}
+	return nil
+}
+
+func blockShouldTriggerNewPoSChallenge(blockHash []byte) bool {
+	bhLen := len(blockHash)
+	return bhLen > 0 && blockHash[bhLen-1]&0x0f == 0
```

Comment:
perhaps a comment here would be nice. this is pretty key logic in terms of frequency of pos challenges, but is subtle